pep-0697.rst

```diff
@@ -0,0 +1,357 @@
+as querying ``__basicsize__`` and plugging it into ``PyType_Spec`` at runtime,
+and divining the correct offset for their extra data.
+This requires making assumptions about memory layout, which the limited API
+is supposed to hide.
+
+
+Extending variable-size objects
+-------------------------------
+
+Another scenario where the traditional way to extend an object does not work
+is variable-sized objects, i.e. ones with non-zero ``tp_itemsize``.
+If the instance struct ends with a variable-length array (such as
+in ``tuple`` or ``int``), subclasses cannot add their own extra data without
+detailed knowledge about how the superclass allocates and uses its memory.
+
+Some types, such as CPython's ``PyHeapType``, handle this by storing
+variable-sized data after the fixed-size struct.
+This means that any subclass can add its own fixed-size data.
+(Only one class in the inheritance hierarchy can use variable-sized data, though.)
+This PEP proposes API that makes this practice easier, and ensures the
+variable-sized data is properly aligned.
+
+Note that many variable-size types, like ``int`` or ``tuple``, do not use
+this mechanism. 
+This PEP does not propose any changes to existing variable-size types (like
+``int`` or ``tuple``) except ``PyHeapType``.
+
+
+Extending ``PyHeapType`` specifically
+-------------------------------------
+
+The motivating problem this PEP solves is creating metaclasses, that is,
+subclasses of ``type``.
+The underlying ``PyHeapTypeObject`` struct is both and variable-sized, and
+opaque in the limited API.
+
+Projects such as language bindings and frameworks that need to attach custom
+data to metaclasses currently resort to questionable workarounds.
+The situation is worse in projects that target the Limited API.
+
+For an example of the currently necessary workarounds, see:
+`nb_type_data_static <https://github.com/wjakob/nanobind/blob/f3044cf44763e105428e4e0cf8f42d951b9cc997/src/nb_type.cpp#L1085>`_
+in the not-yet-released limited-API branch of ``nanobind``
+(a spiritual successor of the popular C++ binding generator ``pybind11``).
+
+
+Rationale
+=========
+
+This PEP proposes a different model: instead of the superclass data being
+part of the subclass data, the extra space a subclass needs is specified
+and accessed separately.
+(How base class data is accessed is left to whomever implements the base class:
+they can for example  provide accessor functions, expose a part of its
+``struct`` for better performance, or do both.)
+
+The proposed mechanism allows using static, read-only ``PyType_Spec``
+even if the superclass struct is opaque, like ``PyTypeObject`` in
+the Limited API.
+
+Combined with a way to create class from ``PyType_Spec`` and a custom metaclass,
+this will allow libraries like nanobind or JPype to create metaclasses
+without making assumptions about ``PyTypeObject``'s memory layout.
+The approach generalizes to non-metaclass types as well.
```

Comment:
But do we care?

Reply:
Yes. It enables the [example in the tutorial](https://docs.python.org/3.10/extending/newtypes_tutorial.html#subclassing-other-types) to work with stable ABI. (And in case you're following Mark Shannon, I should add that *any* stable ABI will need to solve the issue somehow.)

The tutorial's example – extending `list` with an extra C variable – is particularly impractical/simplified, sure, but it shows that currently, subclassing ties you to the ABI of the superclass.

Reply:
Hm. I see no particular problem to extending list, since the list instance itself is not variable-size -- it is a plain struct whose `sizeof` doesn't lie, and the example even works with a static type. (Perhaps that's why list was chosen in the tutorial? :-) Admittedly it doesn't work with the stable ABI (because the typedef `PyListObject` isn't in the ABI) but that's not the same problem you explained to me in the comments above (or I still misunderstood your explanation). The problem with defining types with extra state seems to be caused by the basic type object being variable-sized, not because it's not in the ABI. Or have I got that wrong? The equivalent at the non-metaclass level would seem to me to define a subclass of `tuple` that adds extra state to each tuple instance.

Although now that I read this back I'm not sure whether I've fully understood the problem that you are solving.